File: manji-standard-server-go/pkg/handler/user_handler.go
package handler

import (
	"context"
	"errors"

	"connectrpc.com/connect"
	userv1 "github.com/example/manji-standard-server-go/gen/user/v1"
	"github.com/example/manji-standard-server-go/gen/user/v1/userv1connect"
	"github.com/example/manji-standard-server-go/pkg/domain/entity"
	"github.com/example/manji-standard-server-go/pkg/domain/service"
	"github.com/example/manji-standard-server-go/pkg/usecase"
)

type UserHandler struct {
	userUsecase *usecase.UserUsecase
}

var _ userv1connect.UserServiceHandler = (*UserHandler)(nil)

func NewUserHandler(userUsecase *usecase.UserUsecase) *UserHandler {
	return &UserHandler{userUsecase: userUsecase}
}

func toPbUser(u *entity.User) *userv1.User {
	return &userv1.User{
		Id:            u.ID,
		Email:         u.Email,
		Name:          u.Name,
		CreatedAtUnix: u.CreatedAt.Unix(),
	}
}

func (h *UserHandler) CreateUser(
	ctx context.Context,
	req *connect.Request[userv1.CreateUserRequest],
) (*connect.Response[userv1.CreateUserResponse], error) {
	user, err := h.userUsecase.CreateUser(ctx, req.Msg.GetEmail(), req.Msg.GetName())
	if err != nil {
		if errors.Is(err, service.ErrEmailAlreadyTaken) {
			return nil, connect.NewError(connect.CodeAlreadyExists, err)
		}
		return nil, connect.NewError(connect.CodeInvalidArgument, err)
	}
	return connect.NewResponse(&userv1.CreateUserResponse{User: toPbUser(user)}), nil
}

func (h *UserHandler) GetUser(
	ctx context.Context,
	req *connect.Request[userv1.GetUserRequest],
) (*connect.Response[userv1.GetUserResponse], error) {
	user, err := h.userUsecase.GetUser(ctx, req.Msg.GetId())
	if err != nil {
		return nil, connect.NewError(connect.CodeInternal, err)
	}
	if user == nil {
		return nil, connect.NewError(connect.CodeNotFound, errors.New("user not found"))
	}
	return connect.NewResponse(&userv1.GetUserResponse{User: toPbUser(user)}), nil
}

File: manji-standard-server-ts-hono/src/handler/user-handler.ts
import { Code, ConnectError, type ServiceImpl } from "@connectrpc/connect";
import {
  CreateUserRequest,
  CreateUserResponse,
  GetUserRequest,
  GetUserResponse,
  User as UserPb,
} from "../gen/user/v1/user_pb.js";
import type { UserService as UserServiceDef } from "../gen/user/v1/user_connect.js";
import { EmailAlreadyTakenError } from "../domain/service/user-service.js";
import type { User } from "../domain/entity/user.gen.js";
import type { UserUsecase } from "../usecase/user-usecase.js";

function toPb(user: User): UserPb {
  return new UserPb({
    id: user.id,
    email: user.email,
    name: user.name,
    createdAtUnix: BigInt(Math.floor(user.createdAt.getTime() / 1000)),
  });
}

export function createUserServiceImpl(
  userUsecase: UserUsecase,
): ServiceImpl<typeof UserServiceDef> {
  return {
    async createUser(req: CreateUserRequest): Promise<CreateUserResponse> {
      try {
        const user = await userUsecase.createUser(req.email, req.name);
        return new CreateUserResponse({ user: toPb(user) });
      } catch (err) {
        if (err instanceof EmailAlreadyTakenError) {
          throw new ConnectError(err.message, Code.AlreadyExists);
        }
        const message = err instanceof Error ? err.message : "unknown error";
        throw new ConnectError(message, Code.InvalidArgument);
      }
    },

    async getUser(req: GetUserRequest): Promise<GetUserResponse> {
      const user = await userUsecase.getUser(req.id);
      if (!user) {
        throw new ConnectError("user not found", Code.NotFound);
      }
      return new GetUserResponse({ user: toPb(user) });
    },
  };
}
